### ISSUE TITLE
Three-valued logic instead of exception on comparisons?

### ISSUE THREAD

**Issue by oschulz** [gwater/NumberIntervals.jl#11]
@gwater, at JuliaCon2019, @dpsanders and I discussed possible ways to handle comparisons of intervals.

I was thinking about how to enable generic code with conditional statements, so that it can work both with intervals and standard reals. With the current implementation in `NumberIntervals`, one would need `try`/`catch`, which is not ideal performance-wise.

How about three-valued logic instead? I was thinking about two possible representations

* `true`, `false` and `true_or_false` (with a custom `true_or_false` singleton value of type `TrueOrFalse` similar to `missing` and `Missing`)

* Logic intervals: [true], [false], [false, true]

The latter seems elegant to me (`Bool <: Real` so why not have boolean intervals), but is probably more complex to implement since `IntervalArithmetic` currently doesn't support integer intervals.

Also, using `missing` (or some custom `true_or_false` value) would mimic the current behavior: In a standard `if` statement, it'll result in an exception (`ERROR: TypeError: non-boolean (TrueOrFalse) used in boolean context`). But, unlike the current implementation that throws an exception directly, the user can write code in a fashion that handles `true_or_false`:

```julia
c = a < b
if c == true
    ...
elseif c == false
    ...
else
    ...
end
```

Such code would work generically for with interval numbers and non-interval numbers.

Also, we could overload `ifelse` to handle the `true_or_false` outcome automatically, be returning an interval that spans both `ifelse` return values.

**Comment by gwater** [comment]
Thanks for bringing this up. I see three goals (more or less in that order):
1. API conformity
2. validity
3. performance

Number 1 means I think it's crucial for packages to extend Base functions with correct return types and behaviors (as specified in Base documentation) because it makes combining packages possible in the Julia ecosystem. (@StefanKarpinski made this point in his JuliaCon2019 talk.) For me, that means these functions must return instances of `Boolean` (at least some of the time). Number 2 means, in cases where we cannot make a `true`/`false` judgement, we should not return `true` or `false` and do something else instead.

So, within these constraints; what can we do about Number 3? I understand that exceptions are costly, but only when they happen. So the performance impact depends on the kind of data one expects. The alternative would be a `Union{Boolean,Indeterminate}` return type which has a performance penalty on every call. (Please correct me if that assumption is wrong.) So, I would rather pick the option that empowers users to get the full performance by preprocessing their data. (For example, by checking for overlaps.)

~~Oh, and overloading `ifelse` to return the union in indeterminate cases is a nice idea in any case.~~ EDIT: Of course that only works with return types...

**Comment by oschulz** [comment]
Hi @gwater,

I agree - that's why the boolen intervals solution, while elegant in theory, is probably less practical. But the Union solution (using a `TrueOrFalse` type, or `Indeterminate`) should be both easy to implement and provide full compatibility right? The performance penalty should be low, with Julia's improved handling of small Unions - and the exception comes for free in case people don't handle the `Indeterminate` case.

**Comment by gwater** [comment]
Ok, I looked up the state of union performance, and apparently I missed some significant progress ( see https://julialang.org/blog/2018/08/union-splitting ). I'd like to imagine the hypothetical work-flow for a three-valued return type:
```
ret = iszero(interval)
if isa(ret, Indeterminate)
    # do A
elseif ret
    # do B
else
    # do C
end
```
It doesn't quite roll of the keyboard and fundamentally it means the package doing the check must be aware that NumberIntervals exist (in fact it would have to import the Indeterminate type). (Your example works, too but similarly requires some awareness of this quirk in the interval methods.) I think this breaks a lot of the power of Julia package ecosystem.

In addition, I don't like how unspecific the `TypeError` exception is: Imagine the user applies some 3rd party algorithm to a NumberInterval and wants to catch exceptions due to the indeterminate results. It will be quite difficult since they would have to just assume that all `TypeError`s are caused by NumberIntervals. Under the status quo they could comofortably catch IndeterminateExceptions and be absolutely certain they come from this package.

I wonder if this is actually a usecase of @jrevels' https://github.com/jrevels/Cassette.jl (or if we could implement our own small callback to let end-users define some way of pre-empting frequent indeterminate cases.)

**Comment by gwater** [comment]
It might be useful to return `Base.missing`: [It has a lot of functionality already implemented in Base](https://pkg.julialang.org/docs/julia/THl1k/1.1.1/manual/missing.html) ~~and propagates nicely~~ [unfortunately it causes an exception in typical `&&` and `||` constructs](https://pkg.julialang.org/docs/julia/THl1k/1.1.1/manual/missing.html#Control-Flow-and-Short-Circuiting-Operators-1) . Maybe we could have two modes: *throw exception* or *return missing*. Nevertheless I'm still quite certain the exception is a better default.

**Comment by oschulz** [comment]
My first instinct was to use `Base.missing` as well. But there's a problem: If we want to specialize `ifelse` (and I think that would be really neat), that would be type piracy of the worst kind (involving a very basic function and a very basic data type). That's the main reason why I suggested that `NumberIntervals` should define a custom singleton `trueorfalse` of type `TrueOrFalse`.

> Nevertheless I'm still quite certain the exception is a better default.

Since the `Union` solution would result in an exception too (if the code doesn't use ifelse or take the possible `trueorfalse` outcome into account), is your main worry the generic `TypeError`? After all, it would be something like `ERROR: TypeError: non-boolean (TrueOrFalse) used in boolean context`, so it should be quite obvious to the user that it's caused by `NumberIntervals`.

**Comment by gwater** [comment]
> it should be quite obvious to the user that it's caused by NumberIntervals.

To the actual user, yes, but not to their code. How can they implement exception handling effectively if they cannot distinguish the interval exception from any random TypeError. I really want overcome this limitation in IntervalArithmetic.jl, that every piece of code that is touched by an interval has to be written with awareness for interval quirks.

**Comment by oschulz** [comment]
> To the actual user, yes, but not to their code.

True - on the other hand, they can only profit from a more specialized exception if they write their code in an interval-aware fashion, right?

> I really want overcome this limitation in IntervalArithmetic.jl, that every piece of code that is touched by an interval has to be written with awareness for interval quirks.

But - how? If code is not aware that interval comparison may not return true or false, resp. throw a specific exception, then how can such code (assuming it uses code conditional on such comparisons) run successfully? Maybe I'm misunderstanding something here, though?

**Comment by gwater** [comment]
I'm thinking of use-cases like [this example](https://github.com/gwater/NumberIntervals.jl/blob/master/examples/DifferentialEquationsExample.ipynb). Here, the *user* is aware that they are using intervals but the *package* OrdinaryDiffEq uses no special knowledge about intervals and was implemented without ever looking at NumberIntervals. (Obviously it's not super useful to use intervals for solving differential equations; the point is that it's possible to throw intervals at a random package and it just works.)

**Comment by gwater** [comment]
Basically, we have two use-cases: a) Implementing an algorithm directly on top of intervals with full awareness of the implications and b) implementing an algorithm which puts intervals through some 3rd party code that was written with complete unawareness for intervals. I think it's curcial to get the second case right; the first case can always be treated by the developer in that situation.

**Comment by oschulz** [comment]
Ahh, I get it - in case b), we'd need a very specific exception so that the intervals-aware user of that (not intervals-aware) third-party code can implement a workaround. Yes, that's true - so it would indeed be good to have to modes of operation.

Implementing such a workaround may not be easy though, right? One would have to come up with a sensible result (interval) for it in case where it fails - if the third-party code does something non-trivial,  I guess that may be tricky. Will be application-dependent, of course.

> the first case can always be treated by the developer in that situation

Though having an automatic `ifelse` and the ability to handle true-or-false without catching exceptions (slow) may make that developer's life easier. :-)

**Comment by oschulz** [comment]
Just for fun (to test performance):

```julia
@inline function foo(x)
    if x > 0
        true
    elseif x < 0
        false
    else
        missing
    end
end

A = rand(-100:100, 1024)
R = Array{Union{Bool,Missing}}(undef, size(A)...)

@code_llvm broadcast!(foo, R, A)
```

That broadcast loop actually SIMD-vectorizes now (Julia v1.2-rc2). Handling of small unions has really improved tremendously.

**Comment by oschulz** [comment]
I do see your point regarding interval-unaware code and the need for specific exceptions. I was curious if try-catch is actually a expensive as I feared. Here's a (very simple) benchmark:

```julia
using BenchmarkTools, IntervalArithmetic

# ## Using three-valued logic #####################################

struct TrueOrFalse end
const trueorfalse = TrueOrFalse()

@inline interval_ifelse(c::Bool, a, b) = ifelse(c, a, b)

@inline interval_ifelse(::TrueOrFalse, a, b) = min(a, b)..max(a, b)


@inline function foo1(x)
    if x > 0
        true
    elseif x < 0
        false
    else
        trueorfalse
    end
end

@inline function bar1(x, y)
    interval_ifelse(foo1(x), y, -y)
end


# ## Using exceptions #############################################

struct IndeterminateException <: Exception
    msg
end

@inline function foo2(x)
    if x > 0
        true
    elseif x < 0
        false
    else
        throw(IndeterminateException(x))
    end
end


@inline function bar2(x, y)
    try
        foo2(x) ? y : -y
    catch e
        if e isa IndeterminateException
            min(y,-y)..max(y,-y)
        else
            rethrow()
        end
    end
end


# ## Test data ####################################################

A = rand(-100:100, 1024)
B = rand(-100:100, length(A))
R = Array{Interval{Float64}}(undef, length(A))
```

Unfortunately, exceptions really are very expensive:

```
julia> @benchmark broadcast!(bar1, $R, $A, $B)
BenchmarkTools.Trial: 
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     3.450 μs (0.00% GC)
  median time:      3.610 μs (0.00% GC)
  mean time:        3.636 μs (0.00% GC)
  maximum time:     7.441 μs (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     8

julia> @benchmark broadcast!(bar2, $R, $A, $B)
BenchmarkTools.Trial: 
  memory estimate:  96 bytes
  allocs estimate:  6
  --------------
  minimum time:     227.984 μs (0.00% GC)
  median time:      229.124 μs (0.00% GC)
  mean time:        230.713 μs (0.00% GC)
  maximum time:     338.406 μs (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     1
```

I ran into another problem though, which will make my "automatic ifelse" less attractive - turns out `ifelse` is a builtin function and so can't be specialized. Hence the need for `interval_ifelse` above. Still, such a function would be very helpful when writing interval-aware code.

**Comment by gwater** [comment]
I think #12 implements the kind of behavior you would like. What I find problematic is that it changes the behavior globally. So if two packages both depend on `NumberIntervals` and they each expect a different behavior (return vs exception) then they can never be used at the same time.

**Comment by oschulz** [comment]
Yes, #12 looks exactly like what I meant. One thing: To make optimal use of Julia's optimizations for small unions it would probably be best if `Indeterminate` wouldn't carry any data.

I get your worry about composability if we have both direct exceptions and `Indeterminate` as options. Would it really be so bad if the exception was a `TypeError`, though? Your worry was, from what I understand, that it's only human-readable, not machine-interpretable. But actually it is: The exception you'd get would be something like `TypeError(:if, "", Bool, Indeterminate())`. So code in a `catch` statement could actually determine the specific cause.

**Comment by gwater** [comment]
> it would probably be best if `Indeterminate` wouldn't carry any data.

It doesn't. The arguments are just passed into the exception, otherwise discarded.

Using the `got` field seems to work for `TypeError`:
```julia
       try 
           if iszero(NumberInterval(-1, 1))
               nothing
           end
       catch exc
           if typeof(exc.got) == Indeterminate
                println("hello")
           else
                rethrow(exc)
           end
       end
```
The `==` fallback to `===` should be fixed to throw an exception as well:
```julia
julia> NumberIntervals.Indeterminate(true) == true
false
julia> NumberIntervals.Indeterminate(true) == false
false
```
I wonder if there are more functions that have overly optimistic fallbacks…

**Comment by gwater** [comment]
If we can make returning `Indeterminate()` safe, we could make that the default and maybe keep the exception behavior around for debugging.

**Comment by oschulz** [comment]
I like that approach. Thanks for considering this!

**Comment by gwater** [comment]
@oschulz I've merged #12 , thank you for your suggestion and the helpful discussions. Feel free to open a new issue if there is anything you'd like to see different about the new implementation.

**Comment by oschulz** [comment]
Thanks a lot for all your work on this, @gwater ! I have an idea about how we could use this for parameter space segmentation for likelihood functions. Lot's of fun things to try out. :-)
